nni/retiarii/oneshot/pytorch/sampling.py

```diff
@@ -0,0 +1,339 @@
+        self.ctrl_steps_aggregate = ctrl_steps_aggregate
+        self.grad_clip = grad_clip
+
+    def configure_architecture_optimizers(self):
+        return optim.Adam(self.controller.parameters(), lr=3.5e-4)
+
+    @property
+    def default_replace_dict(self):
+        return {
+            LayerChoice : PathSamplingLayerChoice,
+            InputChoice : PathSamplingInputChoice
+        }
+
+    def training_step(self, batch, batch_idx):
+        # grad manually
+        opts = self.optimizers()
+        if isinstance(opts,list):
+            # list of optimizers
+            # pylint: disable=unsubscriptable-object
+            arc_opt = opts[0]
+            w_opt = opts[1:]
+        else :
+            # no model optimizer
+            arc_opt = opts
+
+        # the batch is composed with x, y, b, where b denote if the data provided
+        # is from the training set
+        batch, is_train_batch = batch
+
+        if is_train_batch:
+            # step 1: train model params
+            self._resample()
+            if w_opt is not None:
+               for opt in w_opt:
+                    opt.zero_grad()
+            w_step_loss = self._extract_user_loss(batch, batch_idx)
+            self.manual_backward(w_step_loss)
+            if w_opt is not None:
+                for opt in w_opt:
+                    opt.step()
+        else:
+            # step 2: train ENAS agent
+            x, y = batch
+            arc_opt.zero_grad()
+            self._resample()
+            with torch.no_grad():
+                logits = self.model(x)
+            breakpoint()
+            # get the default metric of self.model
+            if len(self.model.metrics) == 1:
+                _, metric = next(iter(self.model.metrics.items()))
+            else:
+                if 'default' not in self.model.metrics.keys():
+                    raise KeyError('model.metrics should contain a ''default'' key')
+                metric = self.model.metrics['default']
+
+            reward = metric(logits, y)
+            if self.entropy_weight:
+                reward = reward + self.entropy_weight * self.controller.sample_entropy.item()
+            self.baseline = self.baseline * self.baseline_decay + reward * (1 - self.baseline_decay)
+            rnn_step_loss = self.controller.sample_log_prob * (reward - self.baseline)
+            if self.skip_weight:
+                rnn_step_loss = rnn_step_loss + self.skip_weight * self.controller.sample_skip_penalty
+
+            rnn_step_loss = rnn_step_loss / self.ctrl_steps_aggregate
+            self.manual_backward(rnn_step_loss)
+
+            if (batch_idx + 1) % self.ctrl_steps_aggregate == 0:
+                if self.grad_clip > 0:
+                    nn.utils.clip_grad_norm_(self.controller.parameters(), self.grad_clip)
+                arc_opt.step()
+                arc_opt.zero_grad()
+
+        return self.model.training_step(batch, batch_idx)
+
+    def _resample(self):
```

Comment:
Unify the name. `_resample` or `resample_architecture`?

Even if there is no inheritance to enforce them to be aligned, it's better to have the same style.